### DISCUSSION TITLE
How to run capacitor Android plugin unit tests?

### DISCUSSION THREAD

**Discussion by distante** [ionic-team/capacitor#3605] [Q&A] [answered]
When you creates a new capacitor plugin it comes with a couple of `androidTest` and `test` folder.

I am interested on write and run the unit tests on the `test` folder from the command line (or in a CI) but I do not find any way to do it. And no documentation about exist (or I have not found anything).

There is even an [unanswered Stackoverflow question from Dec 2019.](https://stackoverflow.com/questions/59239101/creating-unit-tests-for-capacitor-native-android-plugin)

How does one run an android unit test on an capacitor plugin?

**Comment by imhoffd** [comment]
Try running `./gradlew test` in the android directory.

**Reply by distante** [reply]
This was easier than I thought. Thank you! 

Coming from just knowing something about Cordova I would like to have this kind of info on the capacitor docs. For people who doesn't know about this like me.

**Reply by imhoffd** [reply]
That's a great idea! I think it would benefit devs to have some concise documentation about the native projects.

**Reply by distante** [reply]
Actually It did not worked as well as I though @dwieeb.

While the task says it was successful It does not check the tests created on `test\src\java\com\getcapacitor\community\myplugin\`.
It always passes even if a set an `assertEquals(true, false);`

So I suppose it need info about where the tests are found?
![image](https://user-images.githubusercontent.com/5490201/94963150-fcd23580-04f7-11eb-946e-4d48b18592cc.png)

Using "run tests on" works (fails) as expected inside Android Studio

![image](https://user-images.githubusercontent.com/5490201/94963198-0fe50580-04f8-11eb-8673-e92b6b9e85f4.png)

I am working here btw: https://github.com/capacitor-community/admob/tree/feat/saninn/add-test-devices__npa-flag

**Reply by distante** [reply]
Extra info, when I change the view to "project" in android studio, it has `node_modules\@capacitor\android\capacitor\src\main\java\com\getcapacitor` as root. No idea why.

![image](https://user-images.githubusercontent.com/5490201/94965460-d31b0d80-04fb-11eb-9491-036736f8f6e1.png)


Edit: Maybe it is related to the `settings.gradle` file?

```
include ':capacitor-android'
project(':capacitor-android').projectDir = new File('../node_modules/@capacitor/android/capacitor')
```

**Comment by distante** [comment]
I am still looking for an answer to this. Maybe somebody with better understanding of the Android configs can help me.

**Reply by hemangsk** [reply]
@distante, I've added android+iOS unit tests for a custom plugin [here](https://github.com/hemangsk/capacitor-mlkit-language/blob/master/android/src/androidTest/java/dev/hemang/mlkit_language/Language/LanguageIdentificationHandlerTest.java) along a CI pipeline for the same.  Here's the [relevant commit](https://github.com/hemangsk/capacitor-mlkit-language/commit/7f24ddd9dfc397242857c5c41902d15ce6646bfb).

**Reply by distante** [reply]
Hi @hemangsk your project has the same problem. I cloned it and changed this in `LanguageIdentificationHandlerTest.java`:
```
    @Test
    public void setLanguageIdentificationOptions() {
        assertEquals(true, false);
    }
```

after running `./gradlew test` or after run the tasks inside android studio the tests are green.

![image](https://user-images.githubusercontent.com/5490201/95500425-16b3c280-09a7-11eb-8c9d-8435fb15eb44.png)

But if I change the `ExampleUnitTest.java` inside `node_modules/@capacitor` it breaks as it should

![image](https://user-images.githubusercontent.com/5490201/95500538-38ad4500-09a7-11eb-9890-4e06c19433c4.png)

@dwieeb maybe can you help us?

**Reply by hemangsk** [reply]
> after running `./gradlew test` or after run the tasks inside android studio the tests are green.
> 
> ![image](https://user-images.githubusercontent.com/5490201/95500425-16b3c280-09a7-11eb-8c9d-8435fb15eb44.png)

Ah, right. Regarding `gradlew test` , turns out it only works for local unit tests, and for instrumented unit tests we will have to do `./gradlew connectedAndroidTest` as it needs emulator / device. 

That being said, I found process for creating and running both these tests types is fairly similar. Here's the result of running the _Tests_ tasks inside android studio, 

1. Running the task in the current state of tests:
<img width="1851" alt="Screenshot 2020-10-09 at 12 49 21 AM" src="https://user-images.githubusercontent.com/13018570/95507908-ca30ad00-09cf-11eb-90a9-ab365b111f98.png">

2. Running tests task after a "destined to fail" assert in `setLanguageIdentificationOptions`.
<img width="1851" alt="Screenshot 2020-10-09 at 12 55 51 AM" src="https://user-images.githubusercontent.com/13018570/95507910-cbfa7080-09cf-11eb-8342-b2e0096a4f3a.png">

I used `Run` > `Run LanguageIdentificationHandlerTest` to execute these instrumented unit tests. For local unit tests, same steps work. 

Similar output with .gradlew script as well:
```
➜  android git:(master) ✗ ./gradlew connectedAndroidTest

> Task :capacitor-android:connectedDebugAndroidTest
Starting 1 tests on Pixel_2_API_R(AVD) - 10

> Task :connectedDebugAndroidTest
Starting 7 tests on Pixel_2_API_R(AVD) - 10

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 9s
88 actionable tasks: 7 executed, 81 up-to-date
```

and on changing, assertNotNull -> assertNull, it fails as expected. (note that I had to comment out a failing assert in the sample `ExampleInstrumentedTests` which were provided by default in the com.getcapacitor.android package.) 
```
➜  android git:(master) ✗ ./gradlew connectedAndroidTest

> Task :capacitor-android:connectedDebugAndroidTest
Starting 1 tests on Pixel_2_API_R(AVD) - 10

> Task :connectedDebugAndroidTest
Starting 7 tests on Pixel_2_API_R(AVD) - 10

dev.hemang.mlkit_language.Language.LanguageIdentificationHandlerTest > setLanguageIdentificationOptions[Pixel_2_API_R(AVD) - 10] FAILED 
        java.lang.AssertionError: expected null, but was:<com.google.mlkit.nl.languageid.LanguageIdentificationOptions@1f>
        at org.junit.Assert.fail(Assert.java:88)

> Task :connectedDebugAndroidTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':connectedDebugAndroidTest'.
> There were failing tests. See the report at: file:///Users/hemang/code/capacitor-mlkit/capacitor-mlkit-language/android/build/reports/androidTests/connected/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 8s
88 actionable tasks: 6 executed, 82 up-to-date
```

**Reply by hemangsk** [reply]
Some update: Tests at https://github.com/capacitor-community/admob/tree/feat/saninn/add-test-devices__npa-flag are working when I tried the same process:

1. 
<img width="1743" alt="Screenshot 2020-10-09 at 2 27 12 AM" src="https://user-images.githubusercontent.com/13018570/95513003-41b60a80-09d7-11eb-8f13-d8b003a0a11a.png">

2. 
<img width="1851" alt="Screenshot 2020-10-09 at 2 25 47 AM" src="https://user-images.githubusercontent.com/13018570/95513018-467abe80-09d7-11eb-9c08-c9a5ff7c987f.png">

**Reply by distante** [reply]
Hi @hemangsk when I said "runned from Android" I meant from here:  
![image](https://user-images.githubusercontent.com/5490201/95551816-c9b60780-0a0b-11eb-960b-0f50e0594fa6.png)

I want to let the test run on a CI, for that I have to call them thru gradlew (please note I do no have any connected tests).

I suppose there is some gradle config that will let us do this but I can not find it.

**Reply by hemangsk** [reply]
Hey @distante , Can you please share the output after running ./gradlew test on the terminal ? The screenshot you shared previously shows the build passing.. but for me, after the builds pass through, it shows the tests execution results.

**Reply by distante** [reply]
It returns @hemangsk :

```
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :capacitor-android:compileReleaseJavaWithJavac
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.

> Task :compileReleaseUnitTestJavaWithJavac
Note: C:\_SSD_DEV\capacitor\capacitor-admob\android\src\test\java\com\getcapacitor\community\admob\executors\AdRewardExecutorTest.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

BUILD SUCCESSFUL in 21s
56 actionable tasks: 41 executed, 15 up-to-date

````

running them from android gradlew task returns more info

```
09:32:56: Executing task 'test'...

Executing tasks: [test] in project C:\_SSD_DEV\capacitor\capacitor-admob\android

Starting Gradle Daemon...
Gradle Daemon started in 1 s 816 ms
> Task :preBuild UP-TO-DATE
> Task :preDebugBuild UP-TO-DATE
> Task :processDebugJavaRes NO-SOURCE
> Task :bundleLibResDebug UP-TO-DATE
> Task :capacitor-android:preBuild UP-TO-DATE
> Task :capacitor-android:preDebugBuild UP-TO-DATE
> Task :capacitor-android:compileDebugAidl NO-SOURCE
> Task :compileDebugAidl NO-SOURCE
> Task :capacitor-android:packageDebugRenderscript NO-SOURCE
> Task :generateDebugBuildConfig UP-TO-DATE
> Task :compileDebugRenderscript NO-SOURCE
> Task :generateDebugResValues UP-TO-DATE
> Task :generateDebugResources UP-TO-DATE
> Task :packageDebugResources UP-TO-DATE
> Task :parseDebugLocalResources UP-TO-DATE
> Task :processDebugManifest UP-TO-DATE
> Task :capacitor-android:generateDebugResValues UP-TO-DATE
> Task :capacitor-android:processDebugManifest UP-TO-DATE
> Task :capacitor-android:compileDebugRenderscript NO-SOURCE
> Task :capacitor-android:generateDebugResources UP-TO-DATE
> Task :capacitor-android:packageDebugResources UP-TO-DATE
> Task :capacitor-android:parseDebugLocalResources UP-TO-DATE
> Task :capacitor-android:generateDebugRFile UP-TO-DATE
> Task :generateDebugRFile UP-TO-DATE
> Task :capacitor-android:generateDebugBuildConfig UP-TO-DATE
> Task :capacitor-android:javaPreCompileDebug UP-TO-DATE
> Task :capacitor-android:compileDebugJavaWithJavac UP-TO-DATE
> Task :capacitor-android:bundleLibCompileDebug UP-TO-DATE
> Task :javaPreCompileDebug UP-TO-DATE
> Task :compileDebugJavaWithJavac UP-TO-DATE
> Task :preDebugUnitTestBuild UP-TO-DATE
> Task :processDebugUnitTestJavaRes NO-SOURCE
> Task :generateDebugUnitTestStubRFile UP-TO-DATE
> Task :javaPreCompileDebugUnitTest UP-TO-DATE
> Task :compileDebugUnitTestJavaWithJavac UP-TO-DATE
> Task :capacitor-android:processDebugJavaRes NO-SOURCE
> Task :capacitor-android:bundleLibResDebug UP-TO-DATE
> Task :capacitor-android:bundleLibRuntimeDebug UP-TO-DATE
> Task :testDebugUnitTest UP-TO-DATE
> Task :preReleaseBuild UP-TO-DATE
> Task :processReleaseJavaRes NO-SOURCE
> Task :bundleLibResRelease UP-TO-DATE
> Task :capacitor-android:preReleaseBuild UP-TO-DATE
> Task :capacitor-android:compileReleaseAidl NO-SOURCE
> Task :compileReleaseAidl NO-SOURCE
> Task :capacitor-android:packageReleaseRenderscript NO-SOURCE
> Task :compileReleaseRenderscript NO-SOURCE
> Task :generateReleaseBuildConfig UP-TO-DATE
> Task :generateReleaseResValues UP-TO-DATE
> Task :generateReleaseResources UP-TO-DATE
> Task :packageReleaseResources UP-TO-DATE
> Task :parseReleaseLocalResources UP-TO-DATE
> Task :processReleaseManifest UP-TO-DATE
> Task :capacitor-android:compileReleaseRenderscript NO-SOURCE
> Task :capacitor-android:generateReleaseResValues UP-TO-DATE
> Task :capacitor-android:generateReleaseResources UP-TO-DATE
> Task :capacitor-android:packageReleaseResources UP-TO-DATE
> Task :capacitor-android:parseReleaseLocalResources UP-TO-DATE
> Task :capacitor-android:processReleaseManifest UP-TO-DATE
> Task :capacitor-android:generateReleaseRFile UP-TO-DATE
> Task :generateReleaseRFile UP-TO-DATE
> Task :capacitor-android:generateReleaseBuildConfig UP-TO-DATE
> Task :capacitor-android:javaPreCompileRelease UP-TO-DATE
> Task :capacitor-android:compileReleaseJavaWithJavac UP-TO-DATE
> Task :capacitor-android:bundleLibCompileRelease UP-TO-DATE
> Task :javaPreCompileRelease UP-TO-DATE
> Task :compileReleaseJavaWithJavac UP-TO-DATE
> Task :preReleaseUnitTestBuild UP-TO-DATE
> Task :generateReleaseUnitTestStubRFile UP-TO-DATE
> Task :javaPreCompileReleaseUnitTest UP-TO-DATE
> Task :compileReleaseUnitTestJavaWithJavac UP-TO-DATE
> Task :processReleaseUnitTestJavaRes NO-SOURCE
> Task :capacitor-android:processReleaseJavaRes NO-SOURCE
> Task :capacitor-android:bundleLibResRelease UP-TO-DATE
> Task :capacitor-android:bundleLibRuntimeRelease UP-TO-DATE
> Task :testReleaseUnitTest UP-TO-DATE
> Task :test UP-TO-DATE
> Task :capacitor-android:preDebugUnitTestBuild UP-TO-DATE
> Task :capacitor-android:generateDebugUnitTestStubRFile UP-TO-DATE
> Task :capacitor-android:javaPreCompileDebugUnitTest UP-TO-DATE
> Task :capacitor-android:compileDebugUnitTestJavaWithJavac UP-TO-DATE
> Task :capacitor-android:processDebugUnitTestJavaRes NO-SOURCE
> Task :capacitor-android:testDebugUnitTest UP-TO-DATE
> Task :capacitor-android:preReleaseUnitTestBuild UP-TO-DATE
> Task :capacitor-android:generateReleaseUnitTestStubRFile UP-TO-DATE
> Task :capacitor-android:javaPreCompileReleaseUnitTest UP-TO-DATE
> Task :capacitor-android:compileReleaseUnitTestJavaWithJavac UP-TO-DATE
> Task :capacitor-android:processReleaseUnitTestJavaRes NO-SOURCE
> Task :capacitor-android:testReleaseUnitTest UP-TO-DATE
> Task :capacitor-android:test UP-TO-DATE

BUILD SUCCESSFUL in 9s
56 actionable tasks: 56 up-to-date
09:33:07: Task execution finished 'test'.

```

**Reply by hemangsk** [reply]
Hey @distante , Thanks for sharing the logs. I also tried `gradlew test` in CLI and was able to reproduce this same output. Its really strange why `gradlew test` doesn't identify those tests in AdMobTest file 🤔  

BTW, I tried to add tests to a fresh plugin and `gradlew test` works there [Action logs](https://github.com/hemangsk/capacitor-tests-example/runs/1235087880#step:7:90) 

[Repo](https://github.com/hemangsk/capacitor-tests-example/)

**Reply by distante** [reply]
Mmmm, I am sure it is a configuration problem. Maybe the plugin template was changed after the AdMob repo was created (I do kot own it).

**Reply by distante** [reply]
Weird enough, I had created a new repo (with the cli) an `gradlew test` finds the tests in 
`com.getcapacitor` but not in `com.mydomain.mypackage.mysub....`  I do not understand it.

**Comment by distante** [comment, marked as answer]
So!!! I finally understood why this was failing.

The problem was that my unit tests were using `JUnit 5` and gradle did not understood how to read them (weirdly enough Android Studio does that just fine)

I had to add this at the top level of my `build.gradle` file:

```
tasks.withType(Test) {
    useJUnitPlatform()
}
```
